virtcontainers/hypervisor.go

```diff
@@ -520,26 +520,12 @@ func (conf *HypervisorConfig) assetPath(t types.AssetType) (string, error) {
 		return a.Path(), nil
 	}
 
-	// We could not find a custom asset for the given type, let's
-	// fall back to the configured ones.
-	switch t {
-	case types.KernelAsset:
-		return conf.KernelPath, nil
-	case types.ImageAsset:
-		return conf.ImagePath, nil
-	case types.InitrdAsset:
-		return conf.InitrdPath, nil
-	case types.HypervisorAsset:
-		return conf.HypervisorPath, nil
-	case types.HypervisorCtlAsset:
-		return conf.HypervisorCtlPath, nil
-	case types.JailerAsset:
-		return conf.JailerPath, nil
-	case types.FirmwareAsset:
-		return conf.FirmwarePath, nil
-	default:
-		return "", fmt.Errorf("Unknown asset type %v", t)
+	assetPath, _, err := t.Annotations()
+	if err != nil {
+		return "", err
 	}
+
+	return assetPath, nil
```

Comment:
This is incorrect as it changes the function to return the asset annotation *name*, not the config *value*. I'm about to push a fix and a new `TestAssetPath()` test to assert the expected behaviour. However, I'm going to have to revert this change meaning we will still have a "list of asset annotations" outside of `asset.go`. This is annoying as I'd like to totally encapsulate that list in `asset.go`.

One possible improvement would be to annotate `HypervisorConfig` with custom struct tags like this:

```go
type HypervisorConfig struct {
    FirmwarePath      string `annotation:"io.katacontainers.config.hypervisor.firmware"`
    HypervisorCtlPath string `annotation:"io.katacontainers.config.hypervisor.ctlpath"`
    HypervisorPath    string `annotation:"io.katacontainers.config.hypervisor.path"`
    ImagePath         string `annotation:"io.katacontainers.config.hypervisor.image"`
    InitrdPath        string `annotation:"io.katacontainers.config.hypervisor.initrd"`
    JailerPath        string `annotation:"io.katacontainers.config.hypervisor.jailer_path"`
    KernelPath        string `annotation:"io.katacontainers.config.hypervisor.kernel"`
}
```

That may make the code clearer and less error prone. However, there would be a runtime impact as we'd need to use `reflect` to map the struct tag (annotation) back to the config value.

Reply:
@jodh-intel I like the idea of encapsulating all the assets constants and logic in assets.go, feel that it is all over the place right now. Maybe using reflection is not such a bad idea if we can cleanup the code and make sure that we do this once in the CreateSandbox code path.
We can address that in a separate PR.

What I did see is that the hypervisor.go has a lot of functions `CustomFirmwareAsset`, `CustomJailerAsset` etc which are not being called at all. 
We need to assess if these functions should be called from `getHypervisorDetails` or we can just get rid of them.